src/collectors/github_runners.rs

```diff
@@ -0,0 +1,207 @@
+use super::default_headers;
+use crate::Config;
+use anyhow::{bail, Result};
+use log::{debug, error, warn};
+use prometheus::core::AtomicI64;
+use prometheus::core::{Desc, GenericGauge};
+use prometheus::proto::MetricFamily;
+use prometheus::{core::Collector, IntGauge, Opts};
+use reqwest::header::{HeaderValue, LINK};
+use reqwest::{Client, Response, StatusCode};
+use std::collections::HashMap;
+use std::sync::{Arc, RwLock};
+use tokio::time::Duration;
+
+const GH_RUNNERS_ENDPOINT: &str =
+    "https://api.github.com/repos/{owner_repo}/actions/runners?per_page=100";
+
+#[derive(Debug, serde::Deserialize)]
+struct ApiResponse {
+    total_count: usize,
+    runners: Vec<Runner>,
+}
+
+#[derive(Debug, serde::Deserialize)]
+struct Runner {
+    id: usize,
+    name: String,
+    os: String,
+    status: String,
+    busy: bool,
+}
+
+#[derive(Clone)]
+pub struct GithubRunners {
+    //api token to use
+    token: String,
+    // repos to track gha runners
+    repos: Vec<String>,
+    // actual metrics
+    metrics: Arc<RwLock<Vec<IntGauge>>>,
+    // default metric description
+    desc: Desc,
+    http: Client,
+}
+
+impl GithubRunners {
+    pub async fn new(config: &Config, http: Client) -> Result<Self> {
+        let token = config.github_token.to_string();
+        let repos: Vec<String> = config
+            .gha_runners_repos
+            .split(',')
+            .map(|v| v.trim().to_string())
+            .collect();
+
+        let rv = Self {
+            token,
+            repos,
+            http,
+            metrics: Arc::new(RwLock::new(Vec::new())),
+            desc: Desc::new(
+                String::from("gha_runner"),
+                String::from("GHA runner's status"),
+                Vec::new(),
+                HashMap::new(),
+            )
+            .unwrap(),
+        };
+
+        let refresh_rate = config.gha_runners_cache_refresh;
+        let mut rv2 = rv.clone();
+        tokio::spawn(async move {
+            loop {
+                if let Err(e) = rv2.update_stats().await {
+                    error!("{:#?}", e);
+                }
+
+                tokio::time::delay_for(Duration::from_secs(refresh_rate)).await;
+            }
+        });
+
+        Ok(rv)
+    }
+
+    async fn update_stats(&mut self) -> Result<()> {
+        let mut gauges = Vec::with_capacity(self.repos.len() * 2);
+        for repo in self.repos.iter() {
+            let mut url: Option<String> = String::from(GH_RUNNERS_ENDPOINT)
+                .replace("{owner_repo}", repo)
+                .into();
+
+            debug!("Updating runner's stats");
+
+            while url.is_some() {
```

Comment:
Minor nit, but this can be replaced with:

```rust
while let Some(url) = url {}
```

...which removes the need for the `.unwrap()` later down. I don't recall if this causes ownership problems off the top of my head, if it does you can replace it with:

```rust
while let Some(url) = url.take() {}
```

Reply:
yeah first suggestion we have ownership issues, went with second option although had to add another variable `endpoint` cos of shadowing of the `url` one or pagination would not work with that while.